pkg/agent/openflow/network_policy.go

```diff
@@ -184,100 +184,202 @@ type conjMatchFlowContext struct {
+	// No need to check if the conjunction ID of the only conjunctive action is the specified ID or not, as it
 	// has been checked in the caller.
 	if len(ctx.actions) == 1 && ctx.flow != nil {
-		if err := ctx.flow.Delete(); err != nil {
-			return err
+		// The expected operation for a "deleted" Openflow entry should be "deletion".
+		return &flowChange{
+			flow:       ctx.flow,
+			changeType: deletion,
 		}
-		ctx.flow = nil
 	} else {
-		// Update Openflow entry with the left conjunctive actions.
+		// Modify the Openflow entry and reset the other conjunctive actions.
 		var actions []*conjunctiveAction
 		for _, act := range ctx.actions {
 			if act.conjID != conjID {
 				actions = append(actions, act)
 			}
 		}
-		err := ctx.installOrUpdateFlow(actions)
-		if err != nil {
-			return err
-		}
+		return ctx.createOrUpdateConjunctiveMatchFlow(actions)
 	}
-	delete(ctx.actions, conjID)
-	return nil
 }
 
-// addAction adds the specified conjunction into conjunctiveMatchFlow's actions, and then updates the conjunctive
-// match flow on the switch. It also installs default drop flow if dropTable is not nil, and the dropFlow is not
-// installed before.
-func (ctx *conjMatchFlowContext) addAction(action *conjunctiveAction) error {
-	// Check if the conjunction exists in conjMatchFlowContext actions or not. If yes, return nil directly.
-	// Otherwise, add the new action, and update the Openflow entry.
+// addAction adds the specified policyRuleConjunction into conjunctiveMatchFlow's actions, and then returns the flowChange.
+func (ctx *conjMatchFlowContext) addAction(action *conjunctiveAction) *flowChange {
+	// Check if the conjunction exists in conjMatchFlowContext actions or not. If yes, return nil immediately.
 	_, found := ctx.actions[action.conjID]
 	if found {
 		return nil
 	}
 
-	// Install or update Openflow entry for the new conjunctiveAction.
-	actions := make([]*conjunctiveAction, 0, len(ctx.actions)+1)
+	// Append current conjunctive action to the existing actions, and then calculate the conjunctive match flow changes.
+	actions := []*conjunctiveAction{action}
 	for _, act := range ctx.actions {
 		actions = append(actions, act)
 	}
-	actions = append(actions, action)
-	err := ctx.installOrUpdateFlow(actions)
-	if err != nil {
-		return err
-	}
-	ctx.actions[action.conjID] = action
-
-	return nil
+	return ctx.createOrUpdateConjunctiveMatchFlow(actions)
 }
 
-func (ctx *conjMatchFlowContext) addDenyAllRule(ruleID uint32) error {
+func (ctx *conjMatchFlowContext) addDenyAllRule(ruleID uint32) {
 	if ctx.denyAllRules == nil {
 		ctx.denyAllRules = make(map[uint32]bool)
 	}
 	ctx.denyAllRules[ruleID] = true
-	return nil
+}
+
+func (ctx *conjMatchFlowContext) delDenyAllRule(ruleID uint32) {
+	// Delete the DENY-ALL rule if it is in context denyAllRules.
+	_, found := ctx.denyAllRules[ruleID]
+	if found {
+		delete(ctx.denyAllRules, ruleID)
+	}
+}
+
+type changeType int
+
+const (
+	add changeType = iota
+	modify
+	deletion
```

Comment:
for consistency this should be "delete"

or alternatively, "insertion", "modification" & "deletion"